### PR TITLE
Aggregations Refactor: Refactor Cardinality Aggregation

### PR DESCRIPTION
**Pull request by colings86**

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java [hunk 1]
```diff
@@ -19,27 +19,58 @@
...
 
-final class CardinalityAggregatorFactory extends ValuesSourceAggregatorFactory.LeafOnly<ValuesSource> {
+public final class CardinalityAggregatorFactory<VS extends ValuesSource> extends ValuesSourceAggregatorFactory.LeafOnly<VS> {
 
-    private final long precisionThreshold;
+    public static final ParseField PRECISION_THRESHOLD_FIELD = new ParseField("precision_threshold");
 
-    CardinalityAggregatorFactory(String name, ValuesSourceParser.Input<ValuesSource> input, long precisionThreshold) {
-        super(name, InternalCardinality.TYPE, input);
+    private long precisionThreshold = -1;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java]: don't other factories use null to represent "unset"?

Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java]: Good point!

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java [hunk 2]
```diff
@@ -51,12 +83,42 @@ protected Aggregator createUnmapped(AggregationContext context, Aggregator paren
...
+    protected ValuesSourceAggregatorFactory<VS> innerReadFrom(String name, ValuesSourceType valuesSourceType,
+            ValueType targetValueType, StreamInput in) throws IOException {
+        CardinalityAggregatorFactory<VS> factory = new CardinalityAggregatorFactory<>(name, valuesSourceType, targetValueType);
+        factory.precisionThreshold = in.readLong();
+        return factory;
+    }
+
+    @Override
+    protected void innerWriteTo(StreamOutput out) throws IOException {
+        out.writeLong(precisionThreshold);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java]: should we handle the null case?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java [hunk 3]
```diff
@@ -51,12 +83,42 @@ protected Aggregator createUnmapped(AggregationContext context, Aggregator paren
...
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        builder.field(PRECISION_THRESHOLD_FIELD.getPreferredName(), precisionThreshold);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/CardinalityAggregatorFactory.java]: only if non null?

**Comment by jpountz** [comment]
LGTM

**Pull request merged**